scipy/stats/_hypotests.py

```diff
@@ -9,6 +9,223 @@
 from ._continuous_distns import chi2, norm
 from scipy.special import gamma, kv, gammaln
 from . import _wilcoxon_data
+import scipy.stats.stats
+from functools import wraps
+from scipy._lib._docscrape import FunctionDoc, Parameter
+import inspect
+
+
+# TODO: add support for `axis` tuples
+def _remove_nans(samples, paired):
+    "Remove nans from paired or unpaired samples"
+    # potential optimization: don't copy arrays that don't contain nans
+    if not paired:
+        return [sample[~np.isnan(sample)] for sample in samples]
+
+    # for paired samples, we need to remove the whole pair when any part
+    # has a nan
+    nans = np.isnan(samples[0])
+    for sample in samples[1:]:
+        nans = nans | np.isnan(sample)
+    not_nans = ~nans
+    return [sample[not_nans] for sample in samples]
+
+
+def _check_empty_inputs(samples, axis):
+    """
+    Check for empty sample; return appropriate output for a vectorized hypotest
+    """
+    # if none of the samples are empty, we need to perform the test
+    if not any((sample.size == 0 for sample in samples)):
+        return None
+    # otherwise, the statistic and p-value will be either empty arrays or
+    # arrays with NaNs. Produce the appropriate array and return it.
+    output_shape = scipy.stats.stats._broadcast_array_shapes(samples, axis)
+    output = np.ones(output_shape) * np.nan
+    return output
+
+
+_name = 'axis'
+_type = "int or None, default: 0"
+_desc = (
+"""If an int, the axis of the input along which to compute the statistic.
+The statistic of each axis-slice (e.g. row) of the input will appear in a
+corresponding element of the output.
+If ``None``, the input will be raveled before computing the statistic."""
+.split('\n'))
+_axis_parameter_doc = Parameter(_name, _type, _desc)
+_axis_parameter = inspect.Parameter(_name,
+                                    inspect.Parameter.KEYWORD_ONLY,
+                                    default=0)
+
+_name = 'nan_policy'
+_type = "{'propagate', 'omit', 'raise'}"
+_desc = (
+"""Defines how to handle input NaNs.
+
+- ``propagate``: if a NaN is present in the axis slice (e.g. row) along
+  which the  statistic is computed, the corresponding entry of the output
+  will be NaN.
+- ``omit``: NaNs will be omitted when performing the calculation.
+  If insufficient data remains in the axis slice along which the
+  statistic is computed, the corresponding entry of the output will be
+  NaN.
+- ``raise``: if a NaN is present, a ``ValueError`` will be raised."""
+.split('\n'))
+_nan_policy_parameter_doc = Parameter(_name, _type, _desc)
+_nan_policy_parameter = inspect.Parameter(_name,
+                                          inspect.Parameter.KEYWORD_ONLY,
+                                          default='propagate')
+
+
+def _vectorize_hypotest_factory(result_object, default_axis=0,
```

Comment:
Should I move this and the preceeding stuff to a separate file? It's 200 lines, and it's used in `stats.py`, `morestate.py`, and `_hypotests.py`.

Reply:
Well some other things that aren't decorators or wrappers belong in there, too. I think the function `_broadcast_shapes` and similar should be there, too. I don't know if we have enough to justify two or more files yet, so is there a less concerning but also less specific name for one?

Reply:
Personally I am not afraid of having more files. Things like `_broadcast_shapes` look actually general enough to move up to a file in `_lib`. In this repo we also have a `decorator`. We could have the same file in here (as this decorator is really targeting stats only).

No strong opinion here though. I could work with `util` if you prefer it.

Reply:
I know we've discussed the possibility of other decorators. How about we call it `util` now and then possibly split decorators into a separate file once we have more than one?

Reply:
I'd prefer to wait then, if we're close here.